raddb/sites-available/coa-relay

```diff
@@ -0,0 +1,289 @@
+#
+#    $ cat <<EOF | radclient 127.0.0.1 coa testing123
+#      User-Name = bob
+#      Cisco-AVPair = "subscriber:sub-qos-policy-out=q_out_uncapped"
+#      EOF
+#
+#  In addition to configuring and activating this site...
+#
+#  Custom attributes must be created in the local dictionary:
+#
+#    ATTRIBUTE  COA-Packet-Type            3010  string
+#    ATTRIBUTE  COA-Packet-DST-IP-Address  3011  ipaddr
+#    ATTRIBUTE  COA-Packet-DST-Port        3012  integer
+#    ATTRIBUTE  COA-Acct-Session-Id        3013  string
+#
+#  A detail writer module must be configured:
+#
+#    detail detail_coa {
+#        filename = ${radacctdir}/detail_coa
+#        escape_filenames = no
+#        permissions = 0600
+#        header = "%t"
+#        locking = yes
+#    }
+#
+
+
+#  Listen on a local CoA port.
+#
+#  This uses the normal set of clients, with the same secret as for
+#  authentication and accounting.
+#
+listen {
+	type = coa
+	ipaddr = 127.0.0.1
+	port = 3799
+	virtual_server = coa
+}
+
+#
+#  Receive CoA/Disconnect, lookup sessions, write them to a detail file
+#
+server coa {
+	#  When a packet is received, it is processed through the
+	#  recv-coa section.  This applies to *both* CoA-Request and
+	#  Disconnect-Request packets.
+	recv-coa {
+
+		# Record the original request type (CoA or Disconnect)
+		#
+		switch Packet-Type {
+			case Disconnect-Request {
+				update request {
+					COA-Packet-Type := "Disconnect-Request"
+				}
+			}
+			case CoA-Request {
+				update request {
+					COA-Packet-Type := "CoA-Request"
+				}
+			}
+		}
+
+		# Lookup all active sessions matching User-Name and/or
+		# Acct-Session-Id and write each session (which includes to
+		# owning NAS and session ID) to a detail file.
+		#
+		# Returns a single result in the format:
+		#
+		# NasIpAddress1#AcctSessionId1|NasIPAddress2#AcctSessionId2|...
+		#
+		update control {
+
+			# Example MySQL lookup
+			Tmp-String-0 := "%{sql:SELECT IFNULL(GROUP_CONCAT(CONCAT(nasipaddress,'#',acctsessionid) separator '|'),'') FROM (SELECT * FROM radacct WHERE ('%{User-Name}'='' OR UserName='%{User-Name}') AND ('%{Acct-Session-Id}'='' OR acctsessionid = '%{Acct-Session-Id}') AND AcctStopTime IS NULL) a}"
+
+			# Example PostgreSQL lookup
+			# Tmp-String-0 := "%{sql:SELECT STRING_AGG(CONCAT(nasipaddress,'#',acctsessionid),'|') FROM (SELECT * FROM radacct WHERE ('%{User-Name}'='' OR UserName='%{User-Name}') AND ('%{Acct-Session-Id}'='' OR acctsessionid = '%{Acct-Session-Id}') AND AcctStopTime IS NULL) a}"
+
+		}
+		if (&control:Tmp-String-0 != "" && "%{explode:&control:Tmp-String-0 |}") {
+			foreach &control:Tmp-String-0 {
+				if ( "%{Foreach-Variable-0}" =~ /([^#]*)#(.*)/ ) {
+					update request {
+
+						COA-Packet-DST-IP-Address := "%{1}"
+						COA-Acct-Session-Id := "%{2}"
+
+						# Set the CoA/Disconnect port
+						COA-Packet-DST-Port := 1700
+
+						# NASs don't usually want to
+						# see User-Name, and
```

Comment:
Some NASes do want to see `User-Name` in a CoA / Disconnect packet.  Those ones will NAK the request if it doesn't contain a `User-Name`

Reply:
I've now left User-Name in with a comment for clearing it.